### PR TITLE
183988 batch job scheduler

### PR DESCRIPTION
**Pull request by satr**

### CODE REVIEW

File: api/secrets.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	secret := corev1.Secret{
+		Type: corev1.SecretTypeOpaque,
+		ObjectMeta: metav1.ObjectMeta{
+			Name: secretName,
+			Labels: map[string]string{
+				kube.RadixAppLabel:       appName,
+				kube.RadixComponentLabel: componentName,
+				kube.RadixJobTypeLabel:   kube.RadixJobTypeJobSchedule,
```
Comment (nilsgstrabo) [inline api/secrets.go:101]: It kube.RadixJobTypeJobSchedule correct, or should it be kube.RadixJobTypeBatchSchedule?

**Review by nilsgstrabo — CHANGES_REQUESTED** [review]

File: api/batches/batch.go [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+)
+
+type batchModel struct {
+	common *api.Model
+}
+
+type Batch interface {
```
[the remark this answers is not in the document]
Comment (satr) [reply api/batches/batch.go:36]: I was not sure about naming and imidlertid called it mode. I agree - handler technically is here. I will rename it

File: api/batches/batch.go [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+	}
+	batchStatus := models.BatchStatus{
+		JobStatus: *jobs.GetJobStatusFromJob(model.common.KubeClient, batch, batchPods),
+	}
+	batchJobsPods, err := model.common.GetPodsForLabelSelector(getLabelSelectorForBatchObjects(batchName))
+	if err != nil {
+		return nil, err
+	}
+	batchJobsPodsMap := getBatchPodsMap(batchJobsPods)
+	batchJobs, err := model.getBatchJobs(batchName)
```
[the remark this answers is not in the document]
Comment (satr) [reply api/batches/batch.go:108]: I'll log this error and return

File: api/batches/batch.go [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+	}
+
+	log.Debugf("Found %v batches for namespace %s", len(allBatchStatuses), model.common.Env.RadixDeploymentNamespace)
+	return allBatchStatuses, nil
+}
+
+//GetBatch Get status of a batch
+func (model *batchModel) GetBatch(batchName string) (*models.BatchStatus, error) {
+	log.Debugf("get batches for namespace: %s", model.common.Env.RadixDeploymentNamespace)
+	batch, err := model.common.GetJob(batchName)
```
[the remark this answers is not in the document]
Comment (satr) [reply api/batches/batch.go:91]: I'll use job-type label for "batch-scheduler"

File: api/batches/batch.go [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+	}
+
+	batchName := generateBatchName(jobComponent)
+
+	descriptionSecret, err := model.common.CreateBatchScheduleDescriptionSecret(batchName, jobComponent, radixDeployment, batchScheduleDescription)
+	if err != nil {
+		return nil, apiErrors.NewFromError(err)
+	}
+
+	if err = model.common.CreateService(batchName, jobComponent, radixDeployment); err != nil {
```
[the remark this answers is not in the document]
Comment (satr) [reply api/batches/batch.go:140]: No, I'll delete it

File: api/jobs.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package api
+
+import (
+	"context"
+
+	batchv1 "k8s.io/api/batch/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+//GetJob Gets a job ba its name
```
[the remark this answers is not in the document]
Comment (satr) [reply api/jobs.go:10]: It uses receiver's properties:
```
func (model *jobModel) getAllJobs() (models.JobList, error) {
	kubeJobs, err := model.common.KubeClient.
		BatchV1().
		Jobs(model.common.Env.RadixDeploymentNamespace).
		List(
			context.TODO(),
			metav1.ListOptions{
				LabelSelector: getLabelSelectorForJobComponent(model.common.Env.RadixComponentName),
```
```
func getLabelSelectorForJobComponent(componentName string) string {
	return labels.SelectorFromSet(map[string]string{
		kube.RadixComponentLabel: componentName,
		kube.RadixJobTypeLabel:   kube.RadixJobTypeJobSchedule,
	}).String()
}
```
I'll rename the methid

File: api/jobs/job.go [hunk 7]
```diff
@@ -0,0 +1,569 @@
...
+
+const (
+	radixJobNameEnvironmentVariable = "RADIX_JOB_NAME"
+)
+
+type jobModel struct {
+	common *api.Model
+}
+
+type Job interface {
```
[the remark this answers is not in the document]
Comment (satr) [reply api/jobs/job.go:37]: Renamed

File: api/jobs.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (satr) [reply api/jobs.go:10]: Method `GetJob` renamed to `GetBatch` (as `GetBatchJob` can be confusing with _getting a job, started by a batch_).
Also used a `getLabelSelectorForJobComponent` to limit labels.

File: api/batches/batch.go [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+		return nil, err
+	}
+	batchJobsPodsMap := getBatchPodsMap(batchJobsPods)
+	batchJobs, err := model.getBatchJobs(batchName)
+	batchStatus.JobStatuses = make([]models.JobStatus, len(batchJobs))
+	for idx, batchJob := range batchJobs {
+		batchStatus.JobStatuses[idx] = *jobs.GetJobStatusFromJob(model.common.KubeClient, batchJob, batchJobsPodsMap[batch.Name])
+	}
+
+	log.Debugf("Found %v jobs for the batche '%s' for namespace '%s'", len(batchJobs), batchName,
```
[the remark this answers is not in the document]
Comment (satr) [reply api/batches/batch.go:114]: Fixed

**Review by Pespiri — APPROVED** [review]

**Pull request merged**